---
  title: "GSEA_benchmark"
author: "Theo Killian"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r, echo = FALSE, results = 'asis'}
suppressPackageStartupMessages(library("dplyr"))
knitr::opts_chunk$set(collapse=TRUE, comment="#>", warning=FALSE, message=FALSE)
```

```{r}
## load libraries
library("ggplot2")
library("tidyverse")
library("dplyr")
```

# GSEA Benchmark

This report describes the benchmarking of various R packages that perform GSEA,
including `goana`, `EnrichmentBrowser` and `GeneTonic`. These packages will be
compared by functionality and performance.

### EnrichmentBrowser

The `EnrichmentBrowser` package implements an analysis pipeline for
high-throughput gene expression data as measured with microarrays and RNA-seq.
In a workflow-like manner, the package brings together a selection of
established Bioconductor packages for gene expression data analysis. It
integrates a wide range of gene set and network enrichment analysis methods and
facilitates combination and exploration of results across methods.

```{r}
library("EnrichmentBrowser")
```

To demonstrate the functionality of the package for RNA-seq data, we consider
transcriptome profiles of four primary human `airway` smooth muscle cell lines
in two conditions: control and treatment with dexamethasone.

```{r}
library("airway")
data(airway)
```

For further analysis, we remove genes with very low read counts and measurements
that are not mapped to an ENSEMBL gene ID.

```{r}
airSE <- airway[grep("^ENSG", rownames(airway)),]
airSE <- airSE[rowMeans(assay(airSE)) > 10,]
# dim(airSE)

assay(airSE)[1:4,1:4]
```

Normalization of high-throughput expression data is essential to make results
within and between experiments comparable. Microarray (intensity measurements)
and RNA-seq (read counts) data typically show distinct features that need to be
normalized for. The function normalize wraps commonly used functionality from
limma for microarray normalization and from EDASeq for RNA-seq normalization.
For specific needs that deviate from these standard normalizations, the user
should always refer to more specific functions/packages. Microarray data is
expected to be single-channel. For two-color arrays, it is expected that
normalization within arrays has been already carried out, e.g. using
normalizeWithinArrays from limma.

```{r}
allSE <- normalize(allSE, norm.method="quantile")
```

The EnrichmentBrowser incorporates established functionality from the limma
package for differential expression analysis between sample groups. This
involves the voom -transformation when applied to RNA-seq data. Alternatively,
differential expression analysis for RNA-seq data can also be carried out based
on the negative binomial distribution with edgeR and DESeq2. 
• GROUP defines the sample groups being contrasted,
• BLOCK defines paired samples or sample blocks, as e.g. for batch effects.
For more information on experimental design, see the limma user guide chapter 9.
For the ALL dataset, the GROUP variable indicates whether the BCR-ABL gene
fusion is present (1) or not (0).

```{r}
allSE$GROUP <- ifelse(allSE$mol.biol == "BCR/ABL", 1, 0)
table(allSE$GROUP)
```

For the airway dataset, it indicates whether the cell lines have been treated
with dexamethasone (1) or not (0).

```{r}
airSE$GROUP <- ifelse(airway$dex == "trt", 1, 0)
table(airSE$GROUP)
```

Paired samples, or in general sample batches/blocks, can be defined via a BLOCK
column in the colData slot. For the airway dataset, the sample blocks correspond
to the four different cell lines.

```{r}
airSE$BLOCK <- airway$cell
table(airSE$BLOCK)
```

For microarray expression data, the deAna function carries out a differential
expression analysis between the two groups based on functionality from the limma
package. Resulting fold changes and t-test derived p-values for each gene are
appended to the rowData slot.

```{r}
allSE <- deAna(allSE, padj.method="BH")
rowData(allSE, use.names=TRUE)
```

Nominal p-values (PVAL) are corrected for multiple testing (ADJ.PVAL) using the
method from Benjamini and Hochberg implemented in the function p.adjust from the
stats package. To get a first overview, we inspect the p-value distribution and
the volcano plot (fold change against p-value).

```{r}
par(mfrow=c(1,2))
pdistr(rowData(allSE)$PVAL)
volcano(rowData(allSE)$FC, rowData(allSE)$ADJ.PVAL)
```

The expression change of highest statistical significance is observed for the
ENTREZ gene 7525.

```{r}
ind.min <- which.min(rowData(allSE)$ADJ.PVAL)
rowData(allSE, use.names=TRUE)[ ind.min, ]
```

This turns out to be the YES proto-oncogene 1 (hsa:7525@KEGG).
For RNA-seq data, the deAna function carries out a differential expression
analysis between the two groups either based on functionality from limma (that
includes the voom transformation), or alternatively, the popular edgeR or DESeq2
package. Here, we use the analysis based on edgeR for demonstration.

```{r}
airSE <- deAna(airSE, de.method="edgeR")
## Excluding 1831 genes not satisfying min.cpm threshold
rowData(airSE, use.names=TRUE)
```

#### ID mapping

Using genomic information from different resources often requires mapping
between different types of gene identifiers. Although primary analysis steps
such as normalization and differential expression analysis can be carried out
independent of the gene ID type, downstream exploration functionality of the
EnrichmentBrowser is consistently based on NCBI Entrez Gene IDs. It is thus, in
this regard, beneficial to initially map gene IDs of a different type to NCBI
Entrez IDs. The function idTypes lists the available ID types for the mapping
depending on the organism under investigation.

```{r}
idTypes("hsa")
```

ID mapping for the airway dataset (from ENSEMBL to ENTREZ gene ids) can then be
carried out using the function idMap .

```{r}
head(rownames(airSE))
```

```{r}
airSE <- idMap(airSE, org="hsa", from="ENSEMBL", to="ENTREZID")
head(rownames(airSE))
```

Now, we subject the ALL and the airway gene expression data to the enrichment
analysis.

#### Enrichment Analysis

```{r}

```

```{r}

```

```{r}

```


### limma goana

Gene Set Enrichment Analysis (GSEA) in this section is performed using
`limma::goana()`. This process requires `ENTREZ` identifiers, which are obtained
by querying the genome wide annotation for the organism of interest, as well as
p-adjusted values. The enriched gene set is subjected to a hypergeometric test
for differential enrichment (DE).

```{r warning=FALSE, message=FALSE}
# library("limma")
# library("GO.db")
# library("org.Mm.eg.db")
# 
# mouse_res$Gene <- as.character(mouse_res$Gene)
# mouse_dat1 <- mouse_res %>%
#               mutate(ENTREZID = mapIds(org.Mm.eg.db, Gene, "ENTREZID", "ENSEMBL") %>%
#               unname())
```

Below shows the number of genes found in the original data.

```{r}
# cat("Dimensions of mouse results ID\n")
# dim(mouse_dat1)[1]
```

To perform GSEA, a *universe* of genes must be constructed. Each gene in the
*universe* must have an `ENTREZ ID` to be testable for differential enrichment.
Below shows the length of the list of such genes in the *universe*.

```{r warning=FALSE, message=FALSE}
# mouse_id1 <- mouse_dat1 %>% filter(!is.na(ENTREZID))
# cat("Dimensions of mouse results with an ENTREZ ID\n")
# dim(mouse_id1)[1]
```

For reference, shown below is the the subset of the data that *does not have* an
`ENTREZ ID` and is not used to construct the universe.

```{r warning=FALSE, message=FALSE}
# mouse_no_id1 <- mouse_dat1 %>% filter(is.na(ENTREZID))
# cat("Dimensions of mouse results without an ENTREZ ID\n")
# dim(mouse_no_id1)[1]
```

`goana` compares the *universe* of genes by looking at all GO terms for "Hs"
(human) with the gene set have provided from the `DESeq2` analysis, returning a
tibble of the GO terms that are significantly differentially enriched after
performing a hypergeometric test. The vector of enriched genes that will be
passed are compared with the universe are those from the `DESeq2` analysis with
a p-adjusted value *lower than 0.05* and an absolute log fold change of
*greater than 1*.

```{r  warning=FALSE, message=FALSE}
# go_goana <- goana(mouse_id1$ENTREZID[(
#                   mouse_id1$padj < 0.05 & abs(mouse_id1$log2FoldChange) > 1)],
#                   mouse_id1$ENTREZID, "Mm") %>% as_tibble()
# 
# go_goana <- go_goana %>% mutate(GOID = mapIds(GO.db, .$Term, "GOID", "TERM") %>%
#                          unname()) %>%
#                          dplyr::select(GOID, everything()) %>% 
#                          arrange(P.DE)
```

#### limma goana results 

Below is a html table displaying the results of significant genes found in the
GSEA analysis for comparison of mouse results.

```{r}
# DT::datatable(go_goana)
```

```{r}
# detach("package:limma", unload = TRUE)
# detach("package:GO.db", unload = TRUE)
# detach("package:org.Mm.eg.db", unload = TRUE)
```

*Session Info*

```{r sessionInfo}
sessionInfo()
```