---
title: "GSEA Benchmark"
author: "Theo Killian"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r, echo = FALSE, results = 'asis'}
suppressPackageStartupMessages(library("dplyr"))
knitr::opts_chunk$set(collapse=TRUE, comment="#>", warning=FALSE, message=FALSE)
```

# Introduction

This report benchmarks popular GSEA packages such as, `EDASeq` and `goseq`,
comparing the results found with each method when analyzing the
`201910_SIGN_STAT2` DE analysis results of *WT TPO vs VEH* data.

```{r load_libraries}
library("goseq")
library("EDASeq")
library("BSgenome.Mmusculus.UCSC.mm10")
library("TxDb.Mmusculus.UCSC.mm10.ensGene")
library("EnrichmentBrowser")
library("DESeq2")
library("dplyr")
# library("geneLenDataBase")
```

We load the results from `201910_SIGN_STAT2` showing the results of DE analysis
of *WT TPO vs VEH*.

```{r load_data}
wt_counts <- readr::read_tsv("./data/wt_counts")
names(wt_counts)[1] <- "ENSEMBL"
head(wt_counts)
```

## EDASeq

We show how to conduct Exploratory Data Analysis (EDA) and normalization for a
typical RNA-Seq experiment using the package `EDASeq`. One can think of EDA for
RNA-Seq as a two-step process: "read-level" EDA helps in discovering lanes with
low sequencing depths, quality issues, and unusual nucleotide frequencies, while
*gene-level* EDA can capture mislabeled lanes, issues with distributional
assumptions (e.g., over-dispersion), and GC-content bias. This package
implements both "within-lane" and "between-lane" normalization procedures, to
account, respectively, for within-lane gene-specific (and possibly
lane-specific) effects on read counts (e.g., related to gene length or
GC-content) and for between-lane distributional differences in read counts
(e.g., sequencing depths).

### Retrieving gene length and GC-content

Two essential features the gene-level EDA normalizes for are gene length and
GC-content. The function `getGeneLengthAndGCContent` is used to retrieve this
information. Given selected ENTREZ or ENSEMBL gene IDs and the organism under
investigation, this can be done either based on BioMart (default) or using BioC
annotation utilities.

```{r getGeneLengthAndGCContent, eval=FALSE}
## NOTE: this function only accepts a column named "ENSEMBL" as its argument
feature <- as.data.frame(getGeneLengthAndGCContent(featureNames(data),
                                                   org="mmu", mode="org.db"))

## unfortunately, not all the genes have a gene length, so these genes, must
## be removed from the analysis, leaving us with 37719 genes to perform DE on
feature <- feature[complete.cases(feature), ]
dim(feature)
```

First, we filter the non-expressed genes, i.e., we consider only the
genes with an average read count greater than 10 across the four lanes
and for which we have length and GC-content information.

```{r filter, message=FALSE, warning=FALSE}
rownames(wt_counts) <- wt_counts$ENSEMBL
wt_counts <- wt_counts %>% dplyr::select(-ENSEMBL)
common <- intersect(rownames(feature), rownames(wt_counts))
length(common)
```

The `EDASeq` package provides the `SeqExpressionSet` class to store gene counts,
(lane-level) information on the sequenced libraries, and (gene-level) feature
information.  We use the data frame `met` created in Section `secRead` for the
lane-level data. As for the feature data, we use gene length and GC-content.

```{r newSeqExpressionSet}
data <- newSeqExpressionSet(counts=as.matrix(wt_counts[common,]),
                            featureData=feature[common,],
                            phenoData=data.frame(
                            conditions=c(rep("KO_IFN", 4), rep("KO_TPO", 4),
                                         rep("KO_VEH", 4), rep("wt_IFN", 4),
                                         rep("wt_TPO", 4), rep("wt_VEH", 4)),
                            row.names=colnames(wt_counts)))
head(counts(data))
```

### Normalization

`EDASeq` considers two main types of effects on gene-level counts: (1)
within-lane gene-specific (and possibly lane-specific) effects, e.g., related to
gene length or GC-content, and (2) effects related to between-lane
distributional differences, e.g., sequencing depth. Accordingly, the functions
`withinLaneNormalization` and `betweenLaneNormalization` adjust for the first
and second type of effects, respectively.  We recommend to normalize for
within-lane effects prior to between-lane normalization. Four within-lane
normalization methods, namely: loess robust local regression of read counts
(log) on a gene feature such as GC-content (`loess`), global-scaling between
feature strata using the median (`median`), global-scaling between feature
strata using the upper-quartile (`upper`), and full-quantile normalization
between feature strata (`full`).

```{r normalization}
# NOTE: it's also possible to normalize for gene length
dataWithin <- withinLaneNormalization(data, "length", which="full")
dataNorm <- betweenLaneNormalization(dataWithin, which="full")

## wt_counts normalized by gene length 
EDASeq_norm_counts <- as.data.frame(dataNorm@assayData$normalizedCounts)
saveRDS(EDASeq_norm_counts, file = "./data/EDASeq_norm_counts.rds")
```

```{r}
boxplot(dataNorm@assayData$normalizedCounts)
```

## goseq

`goseq` will work with any method for determining differential expression (DE).
In order to perform a *GO* analysis on RNA-seq data, `goseq` only requires a
simple named vector, which contains two pieces of information:

1) Measured genes: all genes for which RNA-seq data was gathered for your
experiment. Each element of your vector should be named by a unique gene
identifier.

2) Differentially expressed genes: each element of your vector should be either
a 1 or a 0, where 1 indicates that the gene is differentially expressed and 0
that it is not.

```{r}
supportedOrganisms()
```

```{r}

```

```{r}

```

```{r}

```

```{r}

```

*Session Info*

```{r sessionInfo}
sessionInfo()
```

